### PR TITLE
BUGFIX: Fall back to InClusterConfig

### PR DESCRIPTION
**Pull request by jackkleeman**
Now that we are defaulting to SDS in Pilot instead of secrets as in citadel, this code path is going to be hit a lot more, and I think that it may be a problem that we don't use ClusterConfig. For example, an apiserver doesn't really need to have a cert for kubernetes.default.svc, as most clients  will call it by IP due to ClusterConfig. The problem is essentially that we are making assumptions about what names the apiserver cert will present, and those assumptions differ from the vast vast majority of kubernetes clients. You can reproduce this by running an apiserver that has a cert for its IP only, or a load balancer name, and not for `kubernetes.default.svc`. All clients will work, except for the client in this codepath.

This implementation resolves this issue by attempting to create a client with InClusterConfig if and only if the current implementation does not succeed.

### CODE REVIEW

File: security/pkg/server/ca/server.go [hunk 1]
```diff
@@ -238,9 +240,21 @@ func NewWithGRPC(grpc *grpc.Server, ca CertificateAuthority, ttl time.Duration,
...
+				CAFile: caCertPath,
+			},
+			BearerTokenFile: jwtPath,
+		}
+	}
+
 	// Only add k8s jwt authenticator if SDS is enabled.
 	if sdsEnabled {
-		authenticator, err := authenticate.NewKubeJWTAuthenticator(k8sAPIServerURL, caCertPath, jwtPath,
+		authenticator, err := authenticate.NewKubeJWTAuthenticator(rc.Host+k8sAPIServerPath, rc.CAFile, rc.BearerTokenFile,
```
Comment (JimmyCYJ) [inline security/pkg/server/ca/server.go]: Could you keep the original code path and while adding the new function call? This change might break existing customers when they upgrade their Istio deployment.
Please also add unit tests and integration test to cover the code change.

File: security/pkg/server/ca/server.go [hunk 2]
```diff
@@ -238,9 +240,21 @@ func NewWithGRPC(grpc *grpc.Server, ca CertificateAuthority, ttl time.Duration,
 	authenticators := []authenticator{&authenticate.ClientCertAuthenticator{}}
 	serverCaLog.Info("added client certificate authenticator")
 
+	rc, err := rest.InClusterConfig()
+	if err != nil {
```
Comment (JimmyCYJ) [inline security/pkg/server/ca/server.go]: If this change creates a client which talks to K8S API server, please move this change and K8S API calls into security/pkg/k8s/.

Comment (jackkleeman) [reply security/pkg/server/ca/server.go]: This change doesn't create a new client, it just modifies how we configure `NewKubeJWTAuthenticator`

**Comment by jackkleeman** [comment]
@JimmyCYJ I've added a unit test, I don't think theres any easy way to integration test this given that we would need to modify the k8s control plane configuration so that kubernetes.default.svc is no longer a valid common name for the apiserver. Given that we now only do anything if the normal code path fails, I don't think this is a concern

**Comment by incfly** [comment]
@myidpt @JimmyCYJ @lei-tang can any of you please take a look of this PR? .

**Comment by lei-tang** [comment]
In the issue description, can you add more details of the following items?
1. The problems/bugs of the current implementation.
2. How the proposed implementation solves the problems/bugs of the current implementation?

**Comment by jackkleeman** [comment]
done @lei-tang

**Comment by lei-tang** [comment]
@jackkleeman Thanks for adding more description. 

Quoted from the PR description: "The problem is essentially that we are making assumptions about what names the apiserver cert will present, and those assumptions differ from the vast vast majority of kubernetes clients."

Can you point to the code where "we are making assumptions about what names the apiserver cert will present"? And which part of the PR fixes this problem of making assumptions about what names the apiserver cert will present?

**Comment by jackkleeman** [comment]
@lei-tang 
https://github.com/istio/istio/blob/9629ce39fdecdec4573e96313baeef46702d29f8/security/pkg/server/ca/server.go#L46
The variable `k8sAPIServerURL` is set to `kubernetes.default.svc`. This means that we will initiate an HTTPS connection to that dns name and will thus expect a cert that is signed for that name. 

In this PR, we add the ability to use InClusterConfig https://github.com/istio/istio/blob/9629ce39fdecdec4573e96313baeef46702d29f8/security/pkg/server/ca/server.go#L295
Taking a look at k8s code: https://github.com/kubernetes/client-go/blob/master/rest/config.go#L456 we see that this uses `KUBERNETES_SERVICE_HOST` env var to determine host. This always set in-cluster, and is generally the service clusterIP for the kubernetes apiserver. As a result, we initiate an HTTPS connection directly to the clusterIP and expect a signature for that IP. This is what most client-go clients will do, and therefore its okay to assume that this signature is present. It is not necessarily okay, in my opinion, to assume that the apiserver cert will sign any other names than this one, and there's no benefit to differing from other clients..

**Pull request merged**